File: 0456_132_pattern/find_132_pattern.py
from typing import List


def find132pattern(nums: List[int]) -> bool:
    """
    Time: O(n)
    Space: O(n)
    """
    stack = []
    n3 = float("-inf")
    for num in nums[::-1]:
        if n3 > num:
            return True
        while stack and stack[-1] < num:
            n3 = stack.pop()
        stack.append(num)
    return False

File: 0456_132_pattern/find_132_pattern.java
import java.util.Stack;

class Solution {
    // Time: O(n)
    // Space: O(n)
    public boolean find132pattern(int[] nums) {
        Stack<Integer> stack = new Stack<Integer>();
        int n3 = Integer.MIN_VALUE;
        for (int i = nums.length - 1; i >= 0; i--) {
            if (n3 > nums[i]) {
                return true;
            }
            while (!stack.isEmpty() && stack.peek() < nums[i]) {
                n3 = stack.pop();
            }
            stack.add(nums[i]);
        }
        return false;

    }
}